lib/beaker/hypervisor/abs.rb

```diff
@@ -22,28 +23,90 @@ def provision
       #   "type"     => "centos-7-i386",
       #   "engine"   => "vmpooler",
       # }
-      @resource_hosts.each do |resource_host|
-        type = resource_host['type']
-        type2hosts[type] ||= []
-        type2hosts[type] << resource_host['hostname']
-      end
+      type2hosts = @resource_hosts.group_by { |resource_host| resource_host['type'] }
+
+      missing_templates = []
+      missing_hosts = []
 
       # for each host, get a vm for that template type
       @hosts.each do |host|
         template = host['template']
 
-        raise ArgumentError.new("Failed to provision host '#{host.hostname}' because its 'template' is missing.") if template.nil?
+        if template.nil?
+          missing_templates << host
+          next
+        end
 
         if provisioned_hosts = type2hosts[template]
-          host['vmhostname'] = provisioned_hosts.shift
+          resource_host = provisioned_hosts.shift
+          host['vmhostname'] = resource_host['hostname']
+          if resource_host['engine'] == 'vmpooler'
+            tag(host)
+          end
         else
-          raise ArgumentError.new("Failed to provision host '#{host.hostname}', no template of type '#{host['template']}' was provided.")
+          missing_hosts << host
+        end
+      end
+
+      if host = missing_templates.first
+        raise ArgumentError.new("Failed to provision host '#{host.hostname}' because its 'template' is missing.")
+      end
+
+      if host = missing_hosts.first
+        raise ArgumentError.new("Failed to provision host '#{host.hostname}', no template of type '#{host['template']}' was provided.")
+      end
+
+      type2hosts.each_pair do |_, resource_hosts|
+        if host = resource_hosts.first
+          raise ArgumentError.new("Unexpected host '#{host['hostname']}' of type '#{host['type']}' was provided")
         end
       end
     end
 
     def cleanup
       # nothing to do
     end
+
+    private
+
+    def tag(h)
+      begin
+        uri = URI.parse(@options[:pooling_api] + '/vm/' + h['vmhostname'].split('.')[0])
+
+        http = Net::HTTP.new(uri.host, uri.port)
```

Comment:
I know @tvpartytonight had done recent work on making sure that Beaker's internal http functionality was wrapping Faraday ... makes me think that Beaker has probably had its own internal http functionality for a while.  Do we want/need to call on Net::HTTP, or can we use something built into Beaker?